### PR TITLE
Add Dpbtf2 for computing the (unblocked) Cholesky decomposition of ba…

### PR DESCRIPTION
**Pull request by btracey**
…nded matrices

### CODE REVIEW

**Comment by btracey** [comment]
PTAL @vladimir-ch

File: lapack/gonum/dpbtf2.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/blas"
+	"gonum.org/v1/gonum/blas/blas64"
+)
+
+// Dpbtf2 computes the Cholesky factorization of a symmetric positive banded
+// matrix ab. The matrix ab is n×n with kd diagonal bands. The Choleksy
```
Comment (kortschak) [inline lapack/gonum/dpbtf2.go]: Cholesky

File: lapack/gonum/dpbtf2.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+//  A = L * L^T if ul == blas.Lower
+// ul also specifies the storage of ab. If ul == blas.Upper, then
+// ab is stored as an upper-triangular banded matrix with kd super-diagonals,
+// and if ul == blas.Lower, ab is stored as a lower-triangular banded matrix
+// with kd sub-diagonals. On exit, the banded matrix U or L is stored in-place
+// into ab depending on the value of ul. Dpbtf2 returns whether the factorization
+// was successfully completed.
+//
+// Dpbtf2 is the unblocked version of the algorithm, see Dpbtrf for the blocked
+// version. Dpbtf2 is an internal routine, exported for testing purposes.
```
Comment (kortschak) [inline lapack/gonum/dpbtf2.go]: The internal routine warning should be in its own para.

File: lapack/gonum/dpbtf2.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	"gonum.org/v1/gonum/blas/blas64"
+)
+
+// Dpbtf2 computes the Cholesky factorization of a symmetric positive banded
+// matrix ab. The matrix ab is n×n with kd diagonal bands. The Choleksy
+// factorization computed is
+//  A = U^T * U if ul == blas.Upper
+//  A = L * L^T if ul == blas.Lower
+// ul also specifies the storage of ab. If ul == blas.Upper, then
+// ab is stored as an upper-triangular banded matrix with kd super-diagonals,
```
Comment (kortschak) [inline lapack/gonum/dpbtf2.go]: Given the issues with band matrix storage discussed in the mat PR, it's probably worth writing down the storage with explicit elements, similar to how it's shown in the Fortran.

File: lapack/gonum/dpbtf2.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	}
+	checkSymBanded(ab, n, kd, ldab)
+	if n == 0 {
+		return
+	}
+	bi := blas64.Implementation()
+	kld := max(1, ldab-1)
+	if ul == blas.Upper {
+		for j := 0; j < n; j++ {
+			// Compute U(J,J) and test for non positive-definiteness
```
Comment (kortschak) [inline lapack/gonum/dpbtf2.go]: Period.

File: lapack/gonum/dpbtf2.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+			kn := min(kd, n-j-1)
+			if kn > 0 {
+				bi.Dscal(kn, 1/ajj, ab[j*ldab+1:], 1)
+				bi.Dsyr(blas.Upper, kn, -1, ab[j*ldab+1:], 1, ab[(j+1)*ldab:], kld)
+			}
+		}
+		return true
+	}
+	for j := 0; j < n; j++ {
+		// Compute L(J,J) and test for non positive-definiteness
```
Comment (kortschak) [inline lapack/gonum/dpbtf2.go]: Period.

File: lapack/gonum/general.go [hunk 6]
```diff
@@ -97,6 +97,22 @@ func checkVector(n int, v []float64, inc int) {
...
+	if kd < 0 {
+		panic("lapack: negative bandwidth value")
+	}
+	if lda < kd+1 {
+		panic("lapack: stride less than number of bands")
+	}
+	if len(ab) < (n-1)*lda+kd {
+		panic("lapack: insufficient banded vector length")
+	}
+	return
```
Comment (kortschak) [inline lapack/gonum/general.go]: Delete the `return`.

File: lapack/testlapack/general.go [hunk 7]
```diff
@@ -922,6 +922,131 @@ func equalApproxTriangular(upper bool, n int, a []float64, lda int, b []float64,
...
+	// A matrix is positive definite if and only if it has a Cholesky
+	// decomposition. Generate a random banded lower triangular matrix
+	// to construct the random symmetric matrix.
+	a := make([]float64, n*n)
+	for i := 0; i < n; i++ {
+		for j := max(0, i-kb); j <= i; j++ {
+			a[i*n+j] = rnd.NormFloat64()
+		}
+		a[i*n+i] = math.Abs(a[i*n+i])
+		a[i*n+i] += 1.5 * rnd.Float64() // add to the diagonal to improve the condition number
```
Comment (kortschak) [inline lapack/testlapack/general.go]: Proper sentence.

File: lapack/testlapack/dpbtf2.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+					ldab := kb + 1 + ldoff
+					sym, band := randSymBand(ul, n, ldab, kb, rnd)
+
+					// Compute the Cholesky decomposition of the symmetric matrix
+					ok := impl.Dpotrf(ul, sym.N, sym.Data, sym.Stride)
+					if !ok {
+						panic("bad test: symmetric cholesky decomp failed")
+					}
+
+					// Compute the Cholesky decomposition of the banded matrix
```
Comment (kortschak) [inline lapack/testlapack/dpbtf2.go]: Period.

File: lapack/testlapack/dpbtf2.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	rnd := rand.New(rand.NewSource(1))
+
+	for _, n := range []int{5, 10, 20} {
+		for _, kb := range []int{0, 1, 3, n - 1} {
+			for _, ldoff := range []int{0, 4} {
+				for _, ul := range []blas.Uplo{blas.Upper, blas.Lower} {
+					ldab := kb + 1 + ldoff
+					sym, band := randSymBand(ul, n, ldab, kb, rnd)
+
+					// Compute the Cholesky decomposition of the symmetric matrix
```
Comment (kortschak) [inline lapack/testlapack/dpbtf2.go]: Period.

File: lapack/testlapack/general.go [hunk 10]
```diff
@@ -922,6 +922,131 @@ func equalApproxTriangular(upper bool, n int, a []float64, lda int, b []float64,
...
+	// Construct the SymDense from a*a^T
+	c := make([]float64, n*n)
+	cgen := blas64.General{
+		Rows:   n,
+		Cols:   n,
+		Stride: n,
+		Data:   c,
+	}
+	blas64.Gemm(blas.NoTrans, blas.Trans, 1, agen, agen, 0, cgen)
+
```
Comment (kortschak) [inline lapack/testlapack/general.go]: Delete blank line.

File: lapack/testlapack/general.go [hunk 11]
```diff
@@ -922,6 +922,131 @@ func equalApproxTriangular(upper bool, n int, a []float64, lda int, b []float64,
...
+	sym := blas64.Symmetric{
+		N:      n,
+		Stride: n,
+		Data:   c,
+		Uplo:   ul,
+	}
+
+	b := symToSymBand(ul, c, n, n, kb, ldab)
+
```
Comment (kortschak) [inline lapack/testlapack/general.go]: Delete blank line.

File: lapack/gonum/dpbtf2.go (hunk 1 above)
Comment (btracey) [inline lapack/gonum/dpbtf2.go]: Done.

File: lapack/gonum/dpbtf2.go (hunk 3 above)
Comment (btracey) [inline lapack/gonum/dpbtf2.go]: Done.

I did it fully in the style of Fortran. The documentation with the numbers is fine for `mat`, but I found the indexed version essential for translating the code.

File: lapack/gonum/dpbtf2.go (hunk 2 above)
Comment (btracey) [inline lapack/gonum/dpbtf2.go]: Done.

File: lapack/gonum/dpbtf2.go (hunk 4 above)
Comment (btracey) [inline lapack/gonum/dpbtf2.go]: Done.

File: lapack/gonum/dpbtf2.go (hunk 5 above)
Comment (btracey) [inline lapack/gonum/dpbtf2.go]: Done.

File: lapack/gonum/general.go (hunk 6 above)
Comment (btracey) [inline lapack/gonum/general.go]: Done.

File: lapack/testlapack/dpbtf2.go (hunk 9 above)
Comment (btracey) [inline lapack/testlapack/dpbtf2.go]: Done.

File: lapack/testlapack/dpbtf2.go (hunk 8 above)
Comment (btracey) [inline lapack/testlapack/dpbtf2.go]: Done.

File: lapack/testlapack/general.go (hunk 7 above)
Comment (btracey) [inline lapack/testlapack/general.go]: It is already a sentence, but done :)

File: lapack/testlapack/general.go (hunk 10 above)
Comment (btracey) [inline lapack/testlapack/general.go]: Done.

File: lapack/testlapack/general.go (hunk 11 above)
Comment (btracey) [inline lapack/testlapack/general.go]: Done.

**Pull request merged**